lib/mbedtls/pkcs7_parser.c

```diff
@@ -0,0 +1,506 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * PKCS#7 parser using MbedTLS PKCS#7 library
+ *
+ * Copyright (c) 2024 Linaro Limited
+ * Author: Raymond Mao <raymond.mao@linaro.org>
+ */
+
+#include <log.h>
+#include <linux/kernel.h>
+#include <linux/err.h>
+#include <crypto/public_key.h>
+#include <crypto/pkcs7_parser.h>
+
+static void pkcs7_free_mbedtls_ctx(struct pkcs7_mbedtls_ctx *ctx)
+{
+	if (ctx) {
+		kfree(ctx->content_data);
+		kfree(ctx);
+	}
+}
+
+static void pkcs7_free_sinfo_mbedtls_ctx(struct pkcs7_sinfo_mbedtls_ctx *ctx)
+{
+	if (ctx) {
+		kfree(ctx->authattrs_data);
+		kfree(ctx->content_data_digest);
+		kfree(ctx);
+	}
+}
```

Comment:
Instead of having 2 functions to free the ctx, can't we have one and free everything?
The downside is that the caller has to correctly initialize ctx elements to NULL etc

e.g 
static void pkcs7_free_mbedtls_ctx(struct pkcs7_mbedtls_ctx *ctx)
    kfree(ctx->content_data);
    kfree(ctx->authattrs_data);
    kfree(ctx->content_data_digest);
    kfree(ctx);
}

Reply:
The two context are different, one is `pkcs7_mbedtls_ctx` for a PKCS7 message, the other is `pkcs7_sinfo_mbedtls_ctx` for the signed info structure. We use them in different scenarios: cleanup a PKCS message and cleanup a signed info structure.